### PR TITLE
Refactor Upload/Download: Transfer Abstraction

### PR DESCRIPTION
**Pull request by bgiori**
Previously, each `McuManager` (e.g. FileSystemManager, ImageManager) had separate upload/download implementations. I needed to implement core download and thought it would be a good time to consolidate functionality outside the individual manager classes. 

The new APIs are currently only implemented for CoreDownload which has been tested. If possible I'd like a second pair of eyes on the new APIs, specifically `TransferManager` and `TransferCallable` which could use some work in my opinion.

### `Upload`, `Download` and `Transfer`

Uploads and Downloads share similar properties, mainly data (`byte[]`) and an offset (`int`). These properties have been moved to the `Transfer` abstract class. `Upload` and `Download` abstract classes extend `Transfer` and provide specific handling and parsing for the response. `Upload` and `Download` are extended to implement specific behavior (i.e. packet formatting) and may optionally override callbacks to use custom callback interfaces.

#### CoreDownload Example

Core download is used to download a crash dump from the device.

```java
public class CoreDownload extends Download {
    protected CoreDownload(@NotNull DownloadCallback callback) {
        super(callback);
    }

    @Override
    public DownloadResponse read(int offset) throws McuMgrException {
        return coreLoad(offset);
    }
}
```

### `TransferManager`

* `extends McuManager`

Managers which need to perform data transfers (upload/download) should extend `TransferManager`. `TransferManager` contains a single thread executor which execute transfers and resets the MTU if needed using the `McuManager`s `getMtu()` and `setUploadMtu()` methods. Uploads and downloads are started using `startUpload(Upload)` and `startDownload(Download)` which return an instance to the `TransferController`.

`TransferManager` wraps `TransferCallable` in a runnable which resets the MTU and re-runs the transfer if needed. 

#### Start CoreDownload Example

```java
public TransferController coreDownload(@NotNull DownloadCallback callback) {
    return startDownload(new CoreDownload(callback));
}
```

### `TransferCallable`

* `implements Callable<Transfer>, TransferController`

`TransferCallable` implements the actual transfer, calling the transfer methods and callbacks, and handling calls to `pause()`, `resume()`, and `cancel()`. A `TransferCallable` can only be run once until it completes, fails, or is canceled. When a `TransferCallable` receives an `InsufficientMtuException` from it is thrown by the `call()` method to be handled in the `TransferManager`.

### Other Classes

* **`TransferCallback`**: A generic set of callbacks for transfers. `Transfer` delegates these callbacks to be implemented by `Upload` and `Download` which forward the callbacks to upload/download specific interfaces.
* **`UploadCallback`**: Generic upload callback. Exactly the same as TransferCallback except the methods are renamed.
* **`DownloadCallback`**: Generic download callback. The same as TransferCallback except the methods are renamed and onDownloadCompleted returns the downloaded data (`byte[]`)

### CODE REVIEW

**Comment by philips77** [comment]
Looks very good. Why ImageManager has CoreDownload?

**Comment by bgiori** [comment]
CoreDownload will download a crash dump from a device. I don't think that functionality is implemented within zephyr at the moment though.

File: mcumgr-core/src/main/java/io/runtime/mcumgr/transfer/TransferManager.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                    int mtu = e.getMtu();
+                    if (mMtu == mtu) {
+                        mtu -= 1;
+                    }
+                    boolean isMtuSet = setUploadMtu(mtu);
+                    if (isMtuSet) {
+                        // If the MTU has been set successfully, restart the upload.
+                        transferCallable.getTransfer().reset();
+                        // Rerun the transfer
+                        run();
```
Comment (bgiori) [inline mcumgr-core/src/main/java/io/runtime/mcumgr/transfer/TransferManager.java]: Need to set `mRetry` to false here

**Comment by philips77** [comment]
Is core dump part of Image Manager? There are some core list etc. in another one, afair.

**Comment by bgiori** [comment]
The `coreList()`, `coreErase()` and `coreLoad()` commands are all in the image group (`ImageManager`).

**Comment by bgiori** [comment]
@philips77 This is nearly ready to merge. I still need to set up a zephyr device tomorrow to test all the new transfer implementations.

File: mcumgr-core/src/main/java/io/runtime/mcumgr/managers/FsManager.java [hunk 2]
```diff
@@ -214,37 +303,21 @@ public synchronized void upload(@NotNull String name, @NotNull byte[] data,
...
      */
+    @Deprecated
     public synchronized int getState() {
         return mTransferState;
     }
 
     /**
      * Cancel an undergoing file transfer. Does nothing if no transfer is in progress.
      */
+    @Deprecated
```
Comment (philips77) [inline mcumgr-core/src/main/java/io/runtime/mcumgr/managers/FsManager.java]: How about adding `@deprecated` in comment with a suggestion what to use instead?

File: mcumgr-core/src/main/java/io/runtime/mcumgr/transfer/TransferManager.java (hunk 1 above)
Comment (philips77) [reply mcumgr-core/src/main/java/io/runtime/mcumgr/transfer/TransferManager.java]: Looks like you did set now?

**Comment by bgiori** [comment]
Good call on the comments, got a bit lazy on some of the methods. 

A similar abstraction is planned on iOS.

**Pull request merged**